### PR TITLE
Ledger entry in new RPC framework

### PR DESCRIPTION
**Pull request by cindyyan317**
1 Add currency validator 
2 Change ledger hash validator to uint256 validator which can be used for other fields
3 Add AccountBase58Validator for validate base58 account (not include public key)
4 Add test utility functions to generate ledger entries 
5 Add unittests

### CODE REVIEW

File: src/rpc/ngHandlers/LedgerEntry.h [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        std::optional<uint32_t> ledgerIndex;
+        bool binary = false;
+        // id of this ledger entry: 256 bits hex string
+        std::optional<std::string> index;
+        // index can be extracted from payment_channel, check, escrow, offer
+        // etc, expectedType is used to save the type of index
+        ripple::LedgerEntryType expectedType = ripple::ltANY;
+        // account id to address account root object
+        std::optional<std::string> accountRoot;
+        std::optional<boost::json::object> directory;
```
Comment (godexsoft) [inline src/rpc/ngHandlers/LedgerEntry.h:59]: Is there a way to not leak `json::object` into rpc handler input here? Ideally we want Input to be free from json completely.

File: src/rpc/ngHandlers/LedgerEntry.cpp [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+#include <unordered_map>
+
+namespace RPCng {
+LedgerEntryHandler::Result
+LedgerEntryHandler::process(
+    LedgerEntryHandler::Input input,
+    boost::asio::yield_context& yield) const
+{
+    ripple::uint256 key;
+    if (input.index)
```
Comment (godexsoft) [inline src/rpc/ngHandlers/LedgerEntry.cpp:32]: Maybe this whole thing should be in `tag_invoke` instead? And Input should have a `std::variant` for fields that can be represented in different ways.

**Review by godexsoft — CHANGES_REQUESTED** [review]
Some important design questions to review. Other than that it's pretty good!

File: src/rpc/ngHandlers/LedgerEntry.h (hunk 1 above)
Comment (cindyyan317) [reply src/rpc/ngHandlers/LedgerEntry.h:59]: If we don't use json object here, we will need to create strong type for all the request types, "directory" ,"offer", "escrow" etc. Even we take effort to prepare all the type, for "directory",  we have some check like only one of "owner" and "dir_root" can be available. Current validators can not support this, we can not do it in tag_invoke either.

File: src/rpc/ngHandlers/LedgerEntry.cpp (hunk 2 above)
Comment (cindyyan317) [reply src/rpc/ngHandlers/LedgerEntry.cpp:32]: you mean we calculate the key index in tag_invoke?

File: src/rpc/ngHandlers/LedgerEntry.h (hunk 1 above)
Comment (godexsoft) [reply src/rpc/ngHandlers/LedgerEntry.h:59]: I see. Let's add a `TODO: extract to custom types and remove json usage from Input struct` and move on for now.

File: src/rpc/ngHandlers/LedgerEntry.cpp [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    if (auto const status = std::get_if<RPC::Status>(&lgrInfoOrStatus))
+        return Error{*status};
+
+    auto const lgrInfo = std::get<ripple::LedgerInfo>(lgrInfoOrStatus);
+    auto const ledgerObject =
+        sharedPtrBackend_->fetchLedgerObject(key, lgrInfo.seq, yield);
+    if (!ledgerObject || ledgerObject->size() == 0)
+        return Error{RPC::Status{"entryNotFound"}};
+
+    const ripple::STLedgerEntry sle{
```
Comment (godexsoft) [inline src/rpc/ngHandlers/LedgerEntry.cpp]: West const detected 😃

File: src/rpc/ngHandlers/LedgerEntry.cpp [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+    auto const indexFieldType = std::find_if(
+        indexFieldTypeMap.begin(),
+        indexFieldTypeMap.end(),
+        [&jsonObject](auto const& pair) {
+            auto const& [field, _] = pair;
+            return jsonObject.contains(field) &&
+                jsonObject.at(field).is_string();
+        });
+    if (indexFieldType != indexFieldTypeMap.end())
```
Comment (godexsoft) [inline src/rpc/ngHandlers/LedgerEntry.cpp:244]: minor: I think this whole sheet of `if`s can be made into `if-elseif` with just one `return input` at the very bottom of the function. wdyt?

**Review by godexsoft — CHANGES_REQUESTED** [review]
Few minor comments added, nothing major 🚀

**Review by godexsoft — APPROVED** [review]
LGTM! 🚀

**Pull request merged**